agent/xds/clusters.go

```diff
@@ -21,59 +26,146 @@ func clustersFromSnapshot(cfgSnap *proxycfg.ConfigSnapshot, token string) ([]pro
 	clusters := make([]proto.Message, len(cfgSnap.Proxy.Upstreams)+1)
 
-	clusters[0] = makeAppCluster(cfgSnap)
+	var err error
+	clusters[0], err = makeAppCluster(cfgSnap)
+	if err != nil {
+		return nil, err
+	}
 
 	for idx, upstream := range cfgSnap.Proxy.Upstreams {
-		clusters[idx+1] = makeUpstreamCluster(upstream.Identifier(), cfgSnap)
+		clusters[idx+1], err = makeUpstreamCluster(upstream, cfgSnap)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	return clusters, nil
 }
 
-func makeAppCluster(cfgSnap *proxycfg.ConfigSnapshot) *envoy.Cluster {
-	addr := cfgSnap.Proxy.LocalServiceAddress
-	if addr == "" {
-		addr = "127.0.0.1"
+func makeAppCluster(cfgSnap *proxycfg.ConfigSnapshot) (*envoy.Cluster, error) {
+	var c *envoy.Cluster
+	var err error
+
+	// If we have overriden local cluster config try to parse it into an Envoy cluster
+	if clusterJSONRaw, ok := cfgSnap.Proxy.Config["envoy_local_cluster_json"]; ok {
+		if clusterJSON, ok := clusterJSONRaw.(string); ok {
+			c, err = makeClusterFromUserConfig(clusterJSON)
+			if err != nil {
+				return c, err
+			}
+		}
 	}
-	return &envoy.Cluster{
-		Name: LocalAppClusterName,
-		// TODO(banks): make this configurable from the proxy config
-		ConnectTimeout: 5 * time.Second,
-		Type:           envoy.Cluster_STATIC,
-		// API v2 docs say hosts is deprecated and should use LoadAssignment as
-		// below.. but it doesn't work for tcp_proxy target for some reason.
-		Hosts: []*envoycore.Address{makeAddressPtr(addr, cfgSnap.Proxy.LocalServicePort)},
-		// LoadAssignment: &envoy.ClusterLoadAssignment{
-		//  ClusterName: LocalAppClusterName,
-		//  Endpoints: []endpoint.LocalityLbEndpoints{
-		//    {
-		//      LbEndpoints: []endpoint.LbEndpoint{
-		//        makeEndpoint(LocalAppClusterName,
-		//          addr,
-		//          cfgSnap.Proxy.LocalServicePort),
-		//      },
-		//    },
-		//  },
-		// },
+
+	if c == nil {
+		addr := cfgSnap.Proxy.LocalServiceAddress
+		if addr == "" {
+			addr = "127.0.0.1"
+		}
+		c = &envoy.Cluster{
+			Name:           LocalAppClusterName,
+			ConnectTimeout: 5 * time.Second,
+			Type:           envoy.Cluster_STATIC,
+			// API v2 docs say hosts is deprecated and should use LoadAssignment as
+			// below.. but it doesn't work for tcp_proxy target for some reason.
+			Hosts: []*envoycore.Address{makeAddressPtr(addr, cfgSnap.Proxy.LocalServicePort)},
+			// LoadAssignment: &envoy.ClusterLoadAssignment{
+			//  ClusterName: LocalAppClusterName,
+			//  Endpoints: []endpoint.LocalityLbEndpoints{
+			//    {
+			//      LbEndpoints: []endpoint.LbEndpoint{
+			//        makeEndpoint(LocalAppClusterName,
+			//          addr,
+			//          cfgSnap.Proxy.LocalServicePort),
+			//      },
+			//    },
+			//  },
+			// },
+		}
 	}
+
+	return c, err
 }
 
-func makeUpstreamCluster(name string, cfgSnap *proxycfg.ConfigSnapshot) *envoy.Cluster {
-	return &envoy.Cluster{
-		Name: name,
-		// TODO(banks): make this configurable from the upstream config
```

Comment:
it looks like all the `//TODO(banks)` comments have been lost, is there a pending issue out there as a reminder to make them configurable?

Reply:
Hey sorry about this, this is work in progress, with the code in master you should be able to configure the public listener cluster JSON and the upstream listener JSON.  I plan to do a PR on the docs next week to update to show this configuration.

**Upstream example (see API upstream):**
[](https://github.com/hashicorp/service-mesh-training/blob/master/exercises/lab-04/04-outlier-detection/files/app/ingress.yml)

**Public listener example:**
[](https://github.com/hashicorp/service-mesh-training/blob/master/exercises/lab-04/04-outlier-detection/files/app/cache.yml)

Reply:
Thanks, @nicholasjackson will you be adding any `circuit_breakers` examples to those docs?

Reply:
Note that I'll be enabling default Envoy circuit breaker (i.e. outlier detection) in #5499.

Full support for configuring them without using escape hatch will be in a future release too along with a lot more control of Envoy!